synadm/cli/user.py

```diff
@@ -138,6 +139,55 @@ def deactivate(ctx, helper, user_id, gdpr_erase):
         click.echo("Abort.")
 
 
+@user.command()
+@click.argument("regex", type=str)
+@click.option(
+    "--gdpr-erase", "-e", is_flag=True, default=False,
+    help="""Marks the users as GDPR-erased. This means messages sent by the
+    users will still be visible by anyone that was in the room when these
+    messages were sent, but hidden from other users joining the room
+    afterwards.""", show_default=True)
+@click.option(
+    "--dry-run", "-n", is_flag=True, default=False,
+    help="""Do everything except deactivating users."""
+)
+@click.option(
+    "--batch-size", type=int, default=500, show_default=True,
+    help="""How many users should be requested from the API one at a time.
+    This option has no effect on how many users will be deactivated.
+
+    Increasing this is not necessary in most cases but useful if you have a
+    lot of accounts on your homeserver."""
+)
+@click.pass_obj
+@click.pass_context
+def deactivate_regex(ctx, helper, regex, gdpr_erase, dry_run=False,
+                     batch_size=500):
+    """ Deactivate or GDPR-erase accounts based on regex.
+
+    Does everything normal deactivation does, just for multiple users. The
+    --batch argument may be helpful if you do not want to be prompted to
+    deactivate every single account, and the --dry-run argument is helpful
+    to see which accounts will be deactivated.
+
+    The regex argument takes a string and uses Python's regex matching. It
+    matches based on the full matrix ID."""
+    helper.log.debug(f"Regex: {regex}")
+    # if below fails, turn on debug mode to get the actual given regex.
+    pattern = re.compile(regex)
+    for list_user_response in helper.api.user_list_paginate(batch_size,
+                                                            True, False, "",
+                                                            ""):
+        for user in list_user_response["users"]:
+            if pattern.search(user["name"]):
```

Comment:
I think we have a difference here to the notice send command. See the examples below to see what I mean. 
We've used re.match instead of re.search.

Actually I like the behaviour with your solution in this PR better, but I recall we had some concerns back when we reviewed the notice send feature and thus changed it so it explicitely matches from the beginning (requires the @ or .* to match). At the end of the day we should streamline both commands as good and sensible as possible. Probably better we move in the direction of a behaviour as you suggest here and later on refactor notice send to use what we learnt here. 
If something doesnt work out with the notice send command. we can still change deactivate-regex again. Well, only if we do the refactor soonish, before next release.


```
(synadm38) ~/git/jacksonchen666_synadm (deactivate-regex) $ synadm user deactivate-regex testuser -n
Would deactivate: @testuser1:peek-a-boo.at
Would deactivate: @testuser2:peek-a-boo.at
Would deactivate: @testuser3:peek-a-boo.at
Would deactivate: @testuser4:peek-a-boo.at
Would deactivate: @testuser5:peek-a-boo.at
Would deactivate: @testuser6:peek-a-boo.at
Would deactivate: @testuser:peek-a-boo.at
(synadm38) ~/git/jacksonchen666_synadm (deactivate-regex) $ 
(synadm38) ~/git/jacksonchen666_synadm (deactivate-regex) $ synadm notice send -r @testuser msg 

Note: When sending to a large amount of recipients, consider using the --silent option.

Recipients:
 - @testuser1:peek-a-boo.at
 - @testuser2:peek-a-boo.at
 - @testuser3:peek-a-boo.at
 - @testuser4:peek-a-boo.at
 - @testuser5:peek-a-boo.at
 - @testuser6:peek-a-boo.at
 - @testuser:peek-a-boo.at

Plaintext message:
---
msg
---
Formatted message:
---
msg
---
Send now? [y/N]: 
(synadm38) ~/git/jacksonchen666_synadm (deactivate-regex) $ 
(synadm38) ~/git/jacksonchen666_synadm (deactivate-regex) $ synadm notice send -r testuser msg 

Note: When sending to a large amount of recipients, consider using the --silent option.

Recipients:
(no recipient matched)

Plaintext message:
---
msg
---
Formatted message:
---
msg
---
Send now? [y/N]: 
(synadm38) ~/git/jacksonchen666_synadm (deactivate-regex) $ synadm notice send -r '.*testuser' msg 

Note: When sending to a large amount of recipients, consider using the --silent option.

Recipients:
 - @testuser1:peek-a-boo.at
 - @testuser2:peek-a-boo.at
 - @testuser3:peek-a-boo.at
 - @testuser4:peek-a-boo.at
 - @testuser5:peek-a-boo.at
 - @testuser6:peek-a-boo.at
 - @testuser:peek-a-boo.at

Plaintext message:
---
msg
---
Formatted message:
---
msg
---
Send now? [y/N]: 
(synadm38) ~/git/jacksonchen666_synadm (deactivate-regex) $
```

Reply:
I think having the regex `match` requires the admin to know what they're doing a bit more, whereas using `search` means you can give "spambot" as a regex and have it match based on whether the matrix ID contains "spambot".

`match` could be inconvenient, but `search` might be easy to screw up. Both are regex, so full proper regexes work either way.

Of course, now it's the question of "which one cause we don't want to be inconsistent". I would lean towards `match`.

Search based:
```
> synadm user deactivate-regex -n 42
Would deactivate: @142:localhost:8480
Would deactivate: @242:localhost:8480
Would deactivate: @342:localhost:8480
Would deactivate: @420:localhost:8480
Would deactivate: @421:localhost:8480
Would deactivate: @422:localhost:8480
Would deactivate: @423:localhost:8480
Would deactivate: @424:localhost:8480
Would deactivate: @425:localhost:8480
Would deactivate: @426:localhost:8480
Would deactivate: @427:localhost:8480
Would deactivate: @428:localhost:8480
Would deactivate: @429:localhost:8480
Would deactivate: @42:localhost:8480
Would deactivate: @442:localhost:8480
Would deactivate: @542:localhost:8480
Would deactivate: @642:localhost:8480
> synadm user deactivate-regex -n '@.*42.*?:.*'
Would deactivate: @142:localhost:8480
Would deactivate: @242:localhost:8480
Would deactivate: @342:localhost:8480
Would deactivate: @420:localhost:8480
Would deactivate: @421:localhost:8480
Would deactivate: @422:localhost:8480
Would deactivate: @423:localhost:8480
Would deactivate: @424:localhost:8480
Would deactivate: @425:localhost:8480
Would deactivate: @426:localhost:8480
Would deactivate: @427:localhost:8480
Would deactivate: @428:localhost:8480
Would deactivate: @429:localhost:8480
Would deactivate: @42:localhost:8480
Would deactivate: @442:localhost:8480
Would deactivate: @542:localhost:8480
Would deactivate: @642:localhost:8480
> synadm user deactivate-regex -n @42
Would deactivate: @420:localhost:8480
Would deactivate: @421:localhost:8480
Would deactivate: @422:localhost:8480
Would deactivate: @423:localhost:8480
Would deactivate: @424:localhost:8480
Would deactivate: @425:localhost:8480
Would deactivate: @426:localhost:8480
Would deactivate: @427:localhost:8480
Would deactivate: @428:localhost:8480
Would deactivate: @429:localhost:8480
Would deactivate: @42:localhost:8480
```

Match based:
```
> synadm user deactivate-regex -n 42
> synadm user deactivate-regex -n '@.*42.*?:.*'
Would deactivate: @142:localhost:8480
Would deactivate: @242:localhost:8480
Would deactivate: @342:localhost:8480
Would deactivate: @420:localhost:8480
Would deactivate: @421:localhost:8480
Would deactivate: @422:localhost:8480
Would deactivate: @423:localhost:8480
Would deactivate: @424:localhost:8480
Would deactivate: @425:localhost:8480
Would deactivate: @426:localhost:8480
Would deactivate: @427:localhost:8480
Would deactivate: @428:localhost:8480
Would deactivate: @429:localhost:8480
Would deactivate: @42:localhost:8480
Would deactivate: @442:localhost:8480
Would deactivate: @542:localhost:8480
Would deactivate: @642:localhost:8480
> synadm user deactivate-regex -n @42
Would deactivate: @420:localhost:8480
Would deactivate: @421:localhost:8480
Would deactivate: @422:localhost:8480
Would deactivate: @423:localhost:8480
Would deactivate: @424:localhost:8480
Would deactivate: @425:localhost:8480
Would deactivate: @426:localhost:8480
Would deactivate: @427:localhost:8480
Would deactivate: @428:localhost:8480
Would deactivate: @429:localhost:8480
Would deactivate: @42:localhost:8480
```

Reply:
Additionally, streamlining the output of `deactivate-regex -n` to be similar to sending multiple notices would be kind of... odd, since `deactivate-regex -n` outputs the list of users to deactivate live.

You can see it happening: https://asciinema.org/a/Vj461YV05mmhAof7bUe53y9BQ

(Also, are we putting these messages on stdout? Should be stderr instead cause we definitely do not want to mess with stdout containing human messages requesting interaction when the user expected the stdout to be json or the format they chose)

Reply:
Let's streamline the output of both commands "as good and as sensible as possible". The `notice send` command preview is very special and certainly is a different approach to what `deactivate-regex -n` does. It didn't mean to streamline in that regard! I like how you designed `-n`. Keep it that way please! :-)

Reply:
> (Also, are we putting these messages on stdout? Should be stderr instead cause we definitely do not want to mess with stdout containing human messages requesting interaction when the user expected the stdout to be json or the format they chose)

That is certainly a valid point but up to now we don't have any distinguishing between what output of data and "informational messages" is. What we have though is the `--yes/--batch` global flag. It should always only output data, **as long as it's not errors**.

Certainly sending particular things to stderr instead of stdout would be the right thing to do but that's a whole new chapter. I would like to continue that discussion but I suppose it will be quite some work to refactor all parts of synadm to follow that pattern. But again: Yes it would be the right thing to do for a perfect unix cli tool! :-)

Reply:
Also I see you decided for regex `match`. Ok with me even though I started to like the idea of having the more flexible `search` (I always forget to put @ in the beginning of my regex myself and wonder for a second ;-)). Anyway I think for now `match` is the easier to solution since if we would use `search` here we would need to test `notice send` again, and research why this decision was made. I'm not sure anymore but we talked about it a lot back then. So, yes, let's leave `match`, streamlining is always to be preferred IMHO!